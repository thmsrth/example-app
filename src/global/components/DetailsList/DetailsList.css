.DetailsList__table {
    width: 100%;
    table-layout: auto;
}
