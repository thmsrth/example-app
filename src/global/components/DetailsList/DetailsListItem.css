.DetailsListItem__title {
    padding: var(--section-padding);
    vertical-align: top;
}

.DetailsListItem__value {
    padding: var(--section-padding);
    vertical-align: top;
    text-align: right;
}
