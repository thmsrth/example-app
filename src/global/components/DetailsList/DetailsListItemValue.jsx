import React from 'react';

import PropTypes from 'prop-types';
import keyType from './types/key';

import { formatKeyValue } from '../../utils/FormatHelpers';

const DetailsListItemValue = ({ className, value, keyProp }) => (
  <div className={`DetailsListItemValue ${className}`}>
    {formatKeyValue(value, keyProp)}
  </div>
);

export default DetailsListItemValue;

DetailsListItemValue.propTypes = {
  className: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.func,
    PropTypes.bool,
  ]),
  keyProp: keyType,
};

DetailsListItemValue.defaultProps = {
  className: '',
  value: '',
  keyProp: {},
};
