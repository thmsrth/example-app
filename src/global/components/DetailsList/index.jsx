import React from 'react';
import DetailsList from './DetailsList';

const DetailsListWrapper = ({ ...props }) => (
  <DetailsList
    {...props}
  />
);

export default DetailsListWrapper;
