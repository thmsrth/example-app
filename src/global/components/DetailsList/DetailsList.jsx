import React from 'react';

import PropTypes from 'prop-types';
import keyType from './types/key';

import { bemCls } from '../../utils/ClassNameHelpers';
import {
  getLastKeyValueFromObject,
  getPropertyFromDataItemById,
} from '../../utils/FormatHelpers';

import DetailsListItem from './DetailsListItem';

import './DetailsList.css';

const DetailsList = ({
  item, keys, className, propertyFromData,
}) => (
  <div className={`DetailsList ${bemCls(className, '__list')}`}>
    <table className={bemCls(`DetailsList ${className}`, '__table')}>
      <tbody className={bemCls(`DetailsList ${className}`, '__tbody')}>
        {keys.map((key) => {
          const { propertyFromDataKey } = key;
          const itemData = propertyFromDataKey
            ? getPropertyFromDataItemById(item, propertyFromData, propertyFromDataKey)
            : item;

          const [, lastKeyValue] = getLastKeyValueFromObject(key.name, itemData);
          const value = (lastKeyValue === undefined || lastKeyValue === null) ? '' : lastKeyValue;

          return (
            <DetailsListItem
              key={key.name}
              keyProp={key}
              className={`${bemCls(`DetailsListItem ${className}`, '__property')} ${
                bemCls(`DetailsListItem ${className}`, `__${key.name}`)}`}
              value={value}
            />
          );
        })}
      </tbody>
    </table>
  </div>
);

export default DetailsList;

DetailsList.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  item: PropTypes.object,
  keys: PropTypes.arrayOf(keyType),
  className: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  propertyFromData: PropTypes.object,
};

DetailsList.defaultProps = {
  item: {},
  keys: [],
  className: '',
  propertyFromData: null,
};
