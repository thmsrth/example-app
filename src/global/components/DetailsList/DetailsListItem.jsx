import React from 'react';

import PropTypes from 'prop-types';
import keyType from './types/key';

import { bemCls } from '../../utils/ClassNameHelpers';

import DetailsListItemValue from './DetailsListItemValue';

import './DetailsListItem.css';

const DetailsListItem = ({ keyProp, className, ...otherProps }) => (
  <tr
    className={`DetailsListItem ${className}`}
    tabIndex="0"
  >
    <td className={`${bemCls(`DetailsListItem ${className}`, '__title')}`}>
      {keyProp.title}
    </td>
    <td className={`${bemCls(`DetailsListItem ${className}`, '__value')}`}>
      <DetailsListItemValue
        {...otherProps}
        className={bemCls(className, 'Value')}
        keyProp={keyProp}
      />
    </td>
  </tr>
);

export default DetailsListItem;

DetailsListItem.propTypes = {
  className: PropTypes.string,
  keyProp: keyType,
};

DetailsListItem.defaultProps = {
  className: '',
  keyProp: {},
};
