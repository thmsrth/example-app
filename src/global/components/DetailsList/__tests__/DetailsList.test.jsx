import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import DetailsList from '../DetailsList';

const props = {
  item: {
    id: 1,
    userId: 2,
    name: 'Thomas',
    description: 'huhu',
    number: 1,
    alive: true,
  },
  keys: [
    {
      name: 'name',
      title: 'Name:',
      type: 'text',
    },
  ],
};


describe('DetailsList', () => {
  it('should render the component correctly', () => {
    const tree = shallow(<DetailsList
      {...props}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with property from other data entity by id', () => {
    const tree = shallow(<DetailsList
      {...props}
      propertyFromData={{
        users: [
          {
            id: 2,
            username: 'My User Name',
          },
        ],
      }}
      keys={[
        {
          name: 'user.username',
          title: 'Username:',
          type: 'text',
          propertyFromDataKey: 'users',
        },
      ]}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
