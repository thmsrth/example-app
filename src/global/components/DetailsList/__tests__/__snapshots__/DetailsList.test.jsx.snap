// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsList should render the component correctly 1`] = `
<div
  className="DetailsList __list"
>
  <table
    className="DetailsList__table __table"
  >
    <tbody
      className="DetailsList__tbody __tbody"
    >
      <DetailsListItem
        className="DetailsListItem__property __property DetailsListItem__name __name"
        key="name"
        keyProp={
          Object {
            "name": "name",
            "title": "Name:",
            "type": "text",
          }
        }
        value="Thomas"
      />
    </tbody>
  </table>
</div>
`;

exports[`DetailsList should render the component correctly with property from other data entity by id 1`] = `
<div
  className="DetailsList __list"
>
  <table
    className="DetailsList__table __table"
  >
    <tbody
      className="DetailsList__tbody __tbody"
    >
      <DetailsListItem
        className="DetailsListItem__property __property DetailsListItem__user.username __user.username"
        key="user.username"
        keyProp={
          Object {
            "name": "user.username",
            "propertyFromDataKey": "users",
            "title": "Username:",
            "type": "text",
          }
        }
        value="My User Name"
      />
    </tbody>
  </table>
</div>
`;
